python/hail/dataset.py

```diff
@@ -3425,14 +3425,17 @@ def normalized_genotype_pca(self, k=10, compute_loadings=False, as_array=False):
         :rtype: (list of float, :py:class:`.KeyTable`, :py:class:`.KeyTable`)
         """
 
-        vds = self.annotate_variants_expr('va.mean = gs.map(g => g.GT.gt).sum() / gs.filter(g => isDefined(g.GT)).count()')\
-                .filter_variants_expr('isDefined(va.mean) && gs.filter(g => isDefined(g.GT)).map(g => g.GT.gt).collectAsSet().size() != 1').persist()
+        vds = (self.annotate_variants_expr(
+            'va.AC = gs.map(g => g.GT.gt).sum(), va.nCalled = gs.filter(g => isDefined(g.GT)).count()')
+            .filter_variants_expr('va.AC > 0 && va.AC < 2 * va.nCalled').persist())
         nVariants = vds.count_variants()
         if nVariants == 0:
-            fatal("Cannot run PCA: found 0 variants after filtering out variants with constant genotypes.")
-        print('Running PCA using ' + str(nVariants) + ' variants.')
-        stddev = 'sqrt(va.mean * (2 - va.mean) * ' + str(nVariants) + ' / 2)'
-        result = vds.pca('if (isDefined(g.GT)) (g.GT.gt - va.mean) / ' + stddev + ' else 0', k, compute_loadings, as_array)
+            fatal("Cannot run PCA: found 0 variants after filtering out variants that are all homozygous reference or all homozygous variant.")
+        print('Running PCA using {} variants.'.format(nVariants))
+        result = vds.pca("""let mean = va.AC / va.nCalled in
+            if (isDefined(g.GT)) (g.GT.gt - mean) / sqrt(mean * (2 - mean) * {} / 2)
+            else 0""".format(nVariants),
```

Comment:
Double quotes " should be used for human readable messages.  Single quotes ' should be used for machine readable strings/code.  E.g. I'd change the print to double quotes " and this to triple single quotes.